### PR TITLE
Introduce liblibctest for unikraft

### PR DESCRIPTION
**Pull request by flpostolache**
This is a draft PR that introduces the liblibctest library for Unikraft. Please note that this isn't a final version. Suggestions are welcome.
To run the tests, you need to apply this commit over the unikraft repo to enable downloading a lib as a github repo [1]. Also, if you want to test the lib on musl, you need to disable the myself test suite. This PR enables the disabling of those tests [2].

[1]: https://github.com/unikraft/unikraft/pull/553
[2]: https://github.com/unikraft/unikraft/pull/602

Signed-off-by: Florin Postolache <florin.postolache80@gmail.com>

### CODE REVIEW

File: Config.uk [hunk 1]
```diff
@@ -0,0 +1,13 @@
+menuconfig LIBLIBCTEST
+	bool "Libc-test Unikraft library"
+	default n
```
Comment (razvand) [inline Config.uk:3]:
```suggestion
	default n
	select LIBUKTEST
```

**Review by razvand — CHANGES_REQUESTED** [review]
Add dependency on `LIBUKTEST`.

File: patches/0008-Modify-env-source-to-work-with-uktest.patch [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+-	if (strcmp(s=getenv("TEST"),"3") != 0)
+-		t_error("getenv(\"TEST\"): \"%s\", wanted \"3\"\n", s);
++	UK_TEST_EXPECT_ZERO(setenv("TEST", "3", 1));
++	UK_TEST_EXPECT_ZERO(strcmp(s=getenv("TEST"),"3"));
+ 	/* test failures */
+ 	errno = 0;
+-	if ((r=setenv("","",0)) != -1 || errno != EINVAL)
+-		t_error("setenv(\"\",\"\",0): %d, errno: %d (%s), wanted -1, %d (EINVAL)\n", r, errno, strerror(errno), EINVAL);
+-	return t_status;
++	UK_TEST_ASSERT(((r=setenv("","",0)) != -1 || errno != EINVAL));
```
Comment (DeliaPavel) [inline patches/0008-Modify-env-source-to-work-with-uktest.patch]:
```suggestion
+	UK_TEST_ASSERT(!((r=setenv("","",0)) != -1 || errno != EINVAL));	
```

File: patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
++int f(int pshared, int pi)
+ {
+ 	pthread_t td;
+ 	int r;
+@@ -31,7 +29,7 @@ void f(int pshared, int pi)
+ 	pthread_barrier_t barrier2;
+ 	pthread_mutexattr_t mtx_a;
+ 	pthread_mutex_t mtx;
+-
++	for(int pshared=0; pshared<=1; pshared++) for(int pi=0; pi<=1; pi++){
```
Comment (DeliaPavel) [inline patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch]: Is there any reason for these `for` statements?

File: patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ 
+ 	/* Robust mutexes */
+@@ -63,21 +61,22 @@ void f(int pshared, int pi)
+ 	TEST(r, pthread_mutex_init(&mtx, &mtx_a), "");
+ 	TEST(r, pthread_create(&td, 0, start_wait, (void *[]){ &barrier2, &mtx }), "");
+ 	r = pthread_barrier_wait(&barrier2);
+-	if (r && r != PTHREAD_BARRIER_SERIAL_THREAD)
+-		t_error("pthread_barrier_wait failed: got %d \"%s\", wanted either 0 or %d\n",
+-			r, strerror(r), PTHREAD_BARRIER_SERIAL_THREAD);
++	if(!(r && r != PTHREAD_BARRIER_SERIAL_THREAD))
```
Comment (DeliaPavel) [inline patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch]: Maybe this if statement should keep the previous condition. I am not sure how it should be, but to me, it makes more sense the other way around.
```suggestion
+	if(r && r != PTHREAD_BARRIER_SERIAL_THREAD)
```

File: patches/0096-Modify-strptime-source-to-work-with-uktest.patch [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ 
+ 	/* Glibc */
+-	checkStrptime("1856-07-10", "%F", &tm4);
+-	checkStrptime("683078400", "%s", &tm2);
+-	checkStrptimeTz("+0200", 2, 0);
+-	checkStrptimeTz("-0530", -5, -30);
+-	checkStrptimeTz("-06", -6, 0);
+-
+-	return t_status;
++	UK_TEST_EXPECT_ZERO(checkStrptime("1856-07-10", "%F", &tm4));
```
Comment (DeliaPavel) [inline patches/0096-Modify-strptime-source-to-work-with-uktest.patch]: These tests fail and, judging by the comment, they might be applicable only for glibc.

File: Makefile.uk [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    ifneq ($(filter y,$(CONFIG_LIBLIBCTEST_ACCESS_TESTS) $(CONFIG_LIBUKTEST_ALL)),)
+        LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_BASE)/access_test.c
+    endif
+endif
+
+ifneq ($(filter y,$(CONFIG_LIBLIBCTEST_STRING_TESTS) $(CONFIG_LIBUKTEST_ALL)),)
+	LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_memcpy.c
+    LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_strchr.c
+    LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_strcspn.c
+    LIBLIBCTEST_SRCS-y += $(LIBLIBCTEST_EXTRACTED)/functional/string_strstr.c
```
Comment (DeliaPavel) [inline Makefile.uk]: As far as I could tell, there is no strstr implementation in nolibc so maybe this should be moved in the ifeq as well.

**Comment by kubanrob** [comment]
I had a problem where I wanted to exclude crashing tests from the config. For that I need to check which config option includes that test, looking it up in the Makefiles.
It would be useful to display some kind of name related to the group names displayed in menuconfig. Maybe there could be one testsuite per config option instead of adding all to the same.

File: Config.uk (hunk 1 above)
Comment (flpostolache) [reply Config.uk:3]: Solved

File: patches/0008-Modify-env-source-to-work-with-uktest.patch (hunk 2 above)
Comment (flpostolache) [reply patches/0008-Modify-env-source-to-work-with-uktest.patch]: Solved

File: Makefile.uk (hunk 6 above)
Comment (flpostolache) [reply Makefile.uk]: Moved it.

File: patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch (hunk 4 above)
Comment (flpostolache) [reply patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch]: Changed it.

File: patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch (hunk 3 above)
Comment (flpostolache) [reply patches/0055-Modify-pthread_robust-source-to-work-with-uktest.patch]: No. I removed them.

File: patches/0096-Modify-strptime-source-to-work-with-uktest.patch (hunk 5 above)
Comment (flpostolache) [reply patches/0096-Modify-strptime-source-to-work-with-uktest.patch]: Yes, they are currently not working in `musl`. It is not a problem of Unikraft. @razvand Should they be commented? What do you think?

**Comment by flpostolache** [comment]
@kubanrob I modified the test structure. Now, each test case displays the group name from which is a part of and the name of the source file. If you want to disable the whole group, you can now get the name from the test case label that crashes and simply disable it from the menuconfig. If you want to only disable that specific source, you can get its name form the test case label and comment that line in the Makefile. I think in the future, an option to disable a specific source will be useful.

File: patches/0096-Modify-strptime-source-to-work-with-uktest.patch (hunk 5 above)
Comment (razvand) [reply patches/0096-Modify-strptime-source-to-work-with-uktest.patch]: @maniatro111 , I think they are OK at this point. At some point we would probably do a selection based on libc-type. But I wouldn't make it part of this release.

Having things working (i.e. all tests running, NOT passing) is fine. There shouldn't be crashes when running the tests. Other than that, it's fine.

**Review by DeliaPavel — APPROVED** [review]
Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

**Review by kubanrob — APPROVED** [review]
Thank you @maniatro111

Reviewed-by: Robert Kuban <robert.kuban@opensynergy.com>

**Review by razvand — APPROVED** [review]
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**